crates/m3/tests/merkle_tree.rs

```diff
@@ -0,0 +1,401 @@
+			node_channel.push((self.root_id, self.parent, self.parent_depth, self.parent_index));
+			if self.flush_left {
+				node_channel.pull((
+					self.root_id,
+					self.left,
+					self.parent_depth + 1,
+					2 * self.parent_index,
+				));
+			} else if self.flush_right {
+				node_channel.pull((
+					self.root_id,
+					self.right,
+					self.parent_depth + 1,
+					2 * self.parent_index + 1,
+				));
+			}
+		}
+	}
+
+	impl MerkleRootToken {
+		pub fn fire(
+			&self,
+			node_channel: &mut Channel<NodeFlush>,
+			root_channel: &mut Channel<RootFlush>,
+		) {
+			// Pull the root node value presumed to have been pushed to the nodes channel from the
+			// merkle path table.
+			node_channel.pull((self.root_id, self.digest, 0, 0));
+			// Pull the root node from the roots channel, presumed to have been pushed as a boundary
+			// value.
+			root_channel.pull((self.root_id, self.digest));
+		}
+	}
+
+	impl MerkleTreeTrace {
+		/// Method to generate the trace given the witness values. The function assumes that the
+		/// root_id is the index of the root in the roots vector and that the paths and leaves are
+		/// passed in with their assigned root_id.
+		fn generate(roots: Vec<[u8; 32]>, paths: MerklePath) -> Self {
+			let mut path_vec: Vec<MerklePathToken> = Vec::new();
+			let mut root_vec: Vec<MerkleRootToken> = Vec::new();
+			// Number of times each root is referenced in the paths.
+			let mut root_multiplicities = vec![0; roots.len()];
+
+			for (root_id, index, leaf, path) in paths {
+				root_multiplicities[*root_id as usize] += 1;
+
+				let mut leaf = *leaf;
+				for (i, node) in path.iter().enumerate() {
+					let mut parent = [0u8; 32];
+					if (index >> i) & 1 == 0 {
+						compress(&leaf, node, &mut parent);
+						path_vec.push(MerklePathToken::new(
+							*root_id,
+							leaf,
+							*node,
+							parent,
+							(path.len() - i - 1) as u32,
+							index >> (i + 1),
+							true,
+							false,
+						));
+					} else {
+						compress(node, &leaf, &mut parent);
+						path_vec.push(MerklePathToken::new(
+							*root_id,
+							*node,
+							leaf,
+							parent,
+							(path.len() - i - 1) as u32,
+							index >> (i + 1),
+							false,
+							true,
+						));
+					}
+					leaf = parent;
+				}
+			}
+
+			for (i, root) in roots.iter().enumerate() {
+				for _ in 0..root_multiplicities[i] {
+					root_vec.push(MerkleRootToken::new(i as u8, *root));
+				}
+			}
+
+			Self {
+				nodes: path_vec,
+				root: root_vec,
+			}
+		}
+
+		fn validate(&self, channels: &mut MerkleTreeChannels) {
+			// Push the roots to the roots channel.
+			for root in &self.root {
+				root.fire(&mut channels.nodes, &mut channels.roots);
+			}
+
+			// Push the nodes to the nodes channel.
+			for node in &self.nodes {
+				node.fire(&mut channels.nodes);
+			}
+
+			// Assert that the nodes and roots channels are balanced.
+			channels.nodes.assert_balanced();
+			channels.roots.assert_balanced();
+		}
+	}
+
```

Comment:
What about testing an empty merkle tree? That is a tricky question. Right now I see that it will fail because merkle tree cannot be constructed (`0.is_power_of_two()` would be false). Is this what we want? For the immediate use-case at hand I would say so: we will never have an empty merkle tree and we can always extend this. Or am I missing something?